determinate specification format using the blob content_type

victorperez opened the pull request:
### Summary

Fixes #39921 
Fixes #39983 

Currently the specification class use `format` as `nil` if the blob content_type is a web image, this instructs the variant to retain the source image format which causes some issues under some scenarios where the image format is unknown.

This also cause a bug when a `*.jfif` variant is created using `vips`.

Changes:

* Determinate the specification format using the blob content_type.
* Add tests that are failing before this change and pass after the change.

victorperez committed:
determinate specification format using the blob content_type
Currently the specification class use `format` as `nil` if the blob content_type
is a web image, this instructs the variant to retain the source image format which
causes some issues under some scenarios where the image format is unknown.

This also cause a bug when a `*.jfif` variant is created using `vips`.

Changes:

* Determinate the specification format using the blob content_type.
* Add tests that are failing before this change and pass after the change.

tgxworld commented on activestorage/test/models/variant_test.rb:
```diff
@@ -172,6 +172,25 @@ class ActiveStorage::VariantTest < ActiveSupport::TestCase
+      blob.variant(thumbnail_image: 100).processed
+    end
+  rescue LoadError
+    # libvips not installed
```
Instead of silently passing the test, I think it'll be better to skip with a message here.

victorperez replied:
@tgxworld This is how the current codebase handle it on other `vips` tests but I'm agree with you that skip with a message is better, I will add it to all the tests.

victorperez replied:
@tgxworld Done.

victorperez committed:
skip tests if libvips is not installed
activestorage tests using the vips processor are silently passing if
libvips is not installed, this commit change the behavior to skip the
tests using the exception message instead of silently passing.

jonathanhefner commented on activestorage/app/models/active_storage/variant.rb:
```diff
@@ -108,14 +108,13 @@ def process
             filename: blob.filename,
             content_type: blob.content_type,
-            format: nil
+            format: Rack::Mime::MIME_TYPES.invert[blob.content_type]
```
Since Active Storage uses [MimeMagic](https://github.com/minad/mimemagic) (by way of [Marcel](https://github.com/basecamp/marcel)), you might consider using it here:

```suggestion
            format: MimeMagic.new(blob.content_type).extensions.first
```

victorperez replied:
@jonathanhefner Changed - still digging into the rails codebase so thanks for the suggestion, make sense to use `MimeMagic` since it is already used by Active Storage 👍

jonathanhefner commented on activestorage/test/models/variant_test.rb:
```diff
@@ -172,6 +172,26 @@ class ActiveStorage::VariantTest < ActiveSupport::TestCase
+    ActiveStorage.variant_processor = :vips
+    blob = create_file_blob(filename: "image.jfif", content_type: "image/jpeg")
+
+    assert_nothing_raised do
+      blob.variant(thumbnail_image: 100).processed
+    end
```
Instead of `assert_nothing_raised`, what do you think about asserting on `type` (like the other examples), just to show it works fully as expected?

victorperez replied:
@jonathanhefner Test improved to show that it works as expected.

victorperez committed:
use MimeMagic to determinate specification format
This commit change Rack::Mime::MIME_TYPES to MimeMagic to
determinate the specification format, it also improce the
JFIF variant test to use all asserts as other test to show
it works as expected

georgeclaghorn commented:
> which causes some issues under some scenarios where the image format is unknown

Need some elaboration here. What’s falling back to JPEG for extension-less PNGs and why?

victorperez commented:
@georgeclaghorn Currently `Specification#format` is set as `nil` for web images, it causes the format to be `nil` when `ImageProcessingTransformer#process` is called here: https://github.com/rails/rails/blob/master/activestorage/lib/active_storage/transformers/image_processing_transformer.rb#L10

The process method then use `ImageProcessing::Builder` which define a chainable interface for building processing options and end up calling `ImageProcessing::Pipeline` on the `call` method.

`ImageProcessing::Pipeline` define the default format as `jpg` if no format is known: https://github.com/janko/image_processing/blob/master/lib/image_processing/pipeline.rb#L5

georgeclaghorn commented:
Why is it okay for ImageProcessing to tack a `.jpg` extension onto the destination path? :confused:

If I don’t do that, ImageMagick preserves the original format:

```
$ convert -resize 100x100\> png_image_without_format resized.jpg
$ file --mime-type resized.jpg
/Users/george/Downloads/resized.jpg: image/jpeg

$ convert -resize 100x100\> png_image_without_format resized      
$ file --mime-type ~/Downloads/resized 
/Users/george/Downloads/resized: image/png
```

victorperez commented:
> Why is it okay for ImageProcessing to tack a `.jpg` extension onto the destination path? 😕

Not sure why ImageProcessing use `.jpg` as a default format either, best I could found was the commit message here: https://github.com/janko/image_processing/commit/01f856a29e6c273ca6cbc4f6a7b0aa10499126ac although it does not clarify why `.jpg` is used as a default format.

I can confirm that if the default format is removed from `ImageProcessing::Pipeline` then the variant image maintain the transparency. I can open a PR in the ImageProcessing repository to remove the default format instead of merge this PR, although this PR also fix https://github.com/rails/rails/issues/39921 and I'm not sure if the PR in the ImageProcessing will be accepted since could be a legitimate use case for the default format.

I'm happy to proceed which you consider that is better.

georgeclaghorn commented:
I hadn’t checked VIPS. It turns out that VIPS relies on the destination path’s extension to determine the destination format instead of defaulting to the source format:

```
$ vips rot png_image_without_format rot_png d90
VipsForeignSave: "rot_png" is not a known file format
```

So ImageProcessing does have to add some extension if there isn’t one, and it defaults to JPEG. I’m still a little skeptical that JPEG is a sensible default, but I’ve come around to always passing a target format like you did here.

This highlights that the format determination is duplicated between `ActiveStorage::Variant` (which you’ve updated) and [`ActiveStorage::VariantWithRecord`](https://github.com/rails/rails/blob/3c500511e9e822fb9491f215ac6815272ae1b3c1/activestorage/app/models/active_storage/variant_with_record.rb#L35-L43) (the future replacement for `ActiveStorage::Variant`). I think `ActiveStorage::Blob` should tell `Variant` and `VariantWithRecord` which destination format to use instead of them separately divining it.

victorperez committed:
add destination format to blob
This commit add a `destination_format` method to blob model to tell
`Variant` and `VariantWithRecord` which destination format to use

victorperez commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -154,6 +156,12 @@ def filename
+  # image is created.
+  def destination_format
+    MimeMagic.new(content_type).extensions.first
+  end
```
@georgeclaghorn Should we always use the `content_type` to determinate the destination format or only if the extension is unknown?

georgeclaghorn commented:
Pushed up an alternate take in #40226.
